stac_geoparquet/arrow/_api.py

```diff
@@ -55,14 +104,83 @@ def parse_stac_items_to_arrow(
         )
         return pa.RecordBatchReader.from_batches(schema, batches)
 
-    else:
-        # If schema is _not_ provided, then we must convert to Arrow all at once, or
-        # else it would be possible for a STAC item late in the collection (after the
-        # first chunk) to have a different schema and not match the schema inferred for
-        # the first chunk.
+    elif schema == "FullFile":
         batch = stac_items_to_arrow(items)
+        logger.debug(batch.schema, batch)
         return pa.RecordBatchReader.from_batches(batch.schema, [batch])
 
+    elif schema == "FirstBatch":
+        batches = (
+            stac_items_to_arrow(batch) for batch in batched_iter(items, chunk_size)
+        )
+        return from_batches(batches)
+
+    else:
+        with tempfile.TemporaryDirectory() as tmpdir:
+            logger.info(f"Using temporary directory {tmpdir} for chunked schema")
+            for cnt, chunk in enumerate(batched_iter(items, chunk_size)):
+                batch = stac_items_to_arrow(chunk)
+                if not isinstance(schema, pa.Schema):
+                    schema = batch.schema
+                elif not schema.equals(batch.schema):
+                    logger.info("Unifying schema...")
+                    schema = pa.unify_schemas(
+                        [schema, batch.schema], promote_options="permissive"
+                    )
+                fname = f"{tmpdir}/{cnt}.parquet"
+                to_parquet(
+                    pa.RecordBatchReader.from_batches(batch.schema, [batch]),
+                    output_path=fname,
+                )
+                memlog(f"Batch {cnt}")
+            ds = pa.dataset.dataset(tmpdir, schema=schema, format="parquet")
+            memlog("Created Dataset")
+            batches = ds.to_batches()
```

Comment:
the default batch size for this is 131,072: https://arrow.apache.org/docs/python/generated/pyarrow.dataset.Dataset.html#pyarrow.dataset.Dataset.to_batches

I was still getting some OOM errors so I decided to set this with having batch_size=chunk_size to see if it helps, which I set to 8192 and it resolved the OOM errors for me.